crates/core/transaction/src/effect_hash.rs

```diff
@@ -55,27 +55,26 @@ impl TransactionBody {
     pub fn effect_hash(&self) -> EffectHash {
         let mut state = create_personalized_state(&pbt::TransactionBody::type_url());
 
+        let parameters_hash = self.transaction_parameters.effect_hash();
+        let memo_hash = self
+            .memo
+            .as_ref()
+            .map(|memo| memo.effect_hash())
+            // If the memo is not present, use the all-zero hash to record its absence in
+            // the overall effect hash.
+            .unwrap_or_default();
+        let detection_data_hash = self
+            .detection_data
+            .as_ref()
+            .map(|detection_data| detection_data.effect_hash())
+            // If the detection data is not present, use the all-zero hash to
+            // record its absence in the overall effect hash.
+            .unwrap_or_default();
+
         // Hash the fixed data of the transaction body.
-        state.update(self.transaction_parameters.effect_hash().as_bytes());
-        state.update(self.fee.effect_hash().as_bytes());
-        if self.memo.is_some() {
-            let memo_ciphertext = self.memo.clone();
-            state.update(
-                memo_ciphertext
-                    .expect("memo is some")
-                    .effect_hash()
-                    .as_bytes(),
-            );
-        }
-        if self.detection_data.is_some() {
-            let detection_data = self.detection_data.clone();
-            state.update(
-                detection_data
-                    .expect("detection data is some")
-                    .effect_hash()
-                    .as_bytes(),
-            );
-        }
+        state.update(parameters_hash.as_bytes());
+        state.update(memo_hash.as_bytes());
+        state.update(detection_data_hash.as_bytes());
```

Comment:
cc @redshiftzero ; in the case of a ledger or similar device, we can avoid having the ledger compute FMD clues by passing the detection data hash from the host device to the ledger